### ISSUE TITLE
Make SCM plugin registration more explicit

### ISSUE THREAD

**Issue by mattbrictson** [capistrano/capistrano#1792]
Per @will-in-wi's [comments](https://github.com/capistrano/capistrano/pull/1572/files#r54174714) originally in #1572:

```ruby
# This is somewhat of a hack, because there is no guarantee that a third-
# party SCM will necessarily implement the typical SCM tasks. But it works
# well enough for the built-in SCMs.
def scm_plugin_loaded?
  Rake::Task.tasks.any? { |t| t.name =~ /^[^:]+:create_release/ }
```

> Would it make more sense to require that scm plugins register themselves to avoid this? This feels like it could lead to some difficult to debug situations for a developer creating a new scm plugin. Or possibly throw a warning if fetch(:scm) is set and there are no such tasks.

@mattbrictson:

> Yeah, this is tricky. I guess registration is unavoidable. How should SCM plugins register themselves? Do you think all SCM plugins should inherit from a Plugin::SCM parent class, perhaps? Or add metadata to the plugin that describes what it does, like provides :scm?

@will-in-wi 

> Can we take advantage of the Capistrano::Git.new stanza to have it register itself? IIRC, elsewhere in this change, it tracks things that were called that way…

**Comment by mattbrictson** [comment]
One option is to add a new registration method, `install_scm`. In other words, an SCM would be activated like this:

```ruby
require "capistrano/scm/git"
install_scm Capistrano::SCM::Git
```

It works the same as `install_plugin`, except it sets an extra bit of state that says "an SCM has been registered". The SCMResolver can know for sure that the user has selected an SCM instead of using an error-prone heuristic.

Alternatively, we introduce a `Capistrano::SCM::Plugin` base class that all SCMs must inherit from. That would make it the plugin developer's responsibility to write the plugin correctly, rather than the end-user's responsibility to register it correctly. This solution might be less finicky.

Possible implementation:

```ruby
class Capistrano::Plugin < Rake::TaskLib
  # ...
  def scm?
    @scm
  end
end

class Capistrano::SCM::Plugin < Capistrano::Plugin
  def initialize
    @scm = true
  end
end
```

Thus by inheriting from `Capistrano::SCM::Plugin` you automatically get the `scm? => true` behavior, which is the hint that SCMResovler needs. But at the same time, you don't *have* to inherit as long as you respond to the `scm?` message; any object can be an SCM plugin.

**Comment by will-in-wi** [comment]
This looks good to me. It _might_ make more sense to have the `SCM::Plugin` class instead be a mixin. I try to avoid abstract base classes in Ruby unless there is significant shared code and reused API. In this case, the shared API is very limited.

Alternately, we could define a bit of plugin API with a `#type` method. This could return a symbol like `:scm`.

**Comment by mattbrictson** [comment]
Here's my first take on a PR: #1797

**Comment by mattbrictson** [comment]
Closed via #1797
